File: meshlink-sample/macos/MeshLinkSample/MeshLinkViewModel.swift
// MeshLinkViewModel.swift
// MeshLink macOS Sample — ViewModel managing MeshLink lifecycle and state
//
// Uses `MeshLinkFactory` to create a MeshLink instance backed by
// `IosBleTransport` (CoreBluetooth). Identical API to the iOS sample.

import Foundation
import MeshLink

// MARK: - Peer Info

struct PeerInfo: Identifiable, Equatable {
    let id: String
    var rssi: Int
    var lastSeen: Date
    var firstSeen: Date

    var shortId: String { String(id.prefix(4)) }

    var signalQuality: Double {
        let clamped = min(max(Double(rssi), -100), -30)
        return (clamped + 100) / 70
    }
}

// MARK: - Diagnostic Entry

struct DiagnosticEntry: Identifiable {
    let id: String
    let code: String
    let severity: String
    let timestamp: String
    let payload: String?
}

// MARK: - Config Preset

enum ConfigPreset: String, CaseIterable, Identifiable {
    case chatOptimized = "Chat"
    case fileTransferOptimized = "File Transfer"
    case powerOptimized = "Power Saver"
    case sensorOptimized = "Sensor"

    var id: String { rawValue }

    func makeConfig() -> MeshLinkConfig {
        switch self {
        case .chatOptimized:
            return MeshLinkConfig.companion.chatOptimized { _ in }
        case .fileTransferOptimized:
            return MeshLinkConfig.companion.fileTransferOptimized { _ in }
        case .powerOptimized:
            return MeshLinkConfig.companion.powerOptimized { _ in }
        case .sensorOptimized:
            return MeshLinkConfig.companion.sensorOptimized { _ in }
        }
    }
}

// MARK: - ViewModel

@MainActor
final class MeshLinkViewModel: ObservableObject {

    private var diagnosticCounter = 0

    @Published var isRunning = false
    @Published var logEntries: [String] = []
    @Published var peerCount = 0
    @Published var connectedPeers = 0
    @Published var reachablePeers = 0
    @Published var powerMode = "BALANCED"
    @Published var bufferUsagePercent = 0
    @Published var activeTransfers = 0
    @Published var discoveredPeers: [PeerInfo] = []
    @Published var selectedPeerId: String?
    @Published var currentPreset: ConfigPreset = .chatOptimized
    @Published var currentMtu: Int = 185
    @Published private(set) var maxMessageSize: Int = 10_000
    @Published private(set) var bufferCapacity: Int = 524_288
    @Published var diagnosticEntries: [DiagnosticEntry] = []

    private var meshLink: MeshLink

    init() {
        // Enable BLE debug logging — output appears in Xcode console
        IosBleTransport.companion.debugLogging = true

        let config = ConfigPreset.chatOptimized.makeConfig()
        self.meshLink = MeshLinkFactory.shared.create(config: config)
        self.currentMtu = Int(config.mtu)
        self.maxMessageSize = Int(config.maxMessageSize)
        self.bufferCapacity = Int(config.bufferCapacity)
        startCollectingFlows()
    }

    // MARK: - Configuration

    func applyPreset(_ preset: ConfigPreset) {
        let wasRunning = isRunning
        if wasRunning { stopMesh() }

        currentPreset = preset
        let config = preset.makeConfig()
        meshLink = MeshLinkFactory.shared.create(config: config)
        currentMtu = Int(config.mtu)
        maxMessageSize = Int(config.maxMessageSize)
        bufferCapacity = Int(config.bufferCapacity)
        startCollectingFlows()
        log("⚙️ Applied preset: \(preset.rawValue)")

        if wasRunning { startMesh() }
    }

    func resetToDefaults() {
        applyPreset(.chatOptimized)
        log("⚙️ Reset to defaults")
    }

    // MARK: - Peer Selection

    func selectPeer(_ peerId: String?) {
        selectedPeerId = peerId
    }

    func peerDetail(_ peerIdHex: String) -> PeerDetail? {
        return meshLink.peerDetail(peerIdHex: peerIdHex)
    }

    // MARK: - Mesh Lifecycle

    func startMesh() {
        let result = meshLink.start()
        guard result != nil else {
            log("❌ Start failed — check Bluetooth is enabled and permissions are granted")
            return
        }
        isRunning = true
        log("🟢 Mesh started — scanning for peers via BLE")
    }

    func stopMesh() {
        meshLink.stop()
        isRunning = false
        discoveredPeers.removeAll()
        peerCount = 0
        connectedPeers = 0
        reachablePeers = 0
        activeTransfers = 0
        bufferUsagePercent = 0
        powerMode = "BALANCED"
        log("🔴 Mesh stopped")
    }

    // MARK: - Messaging

    func sendMessage(to recipientHex: String, message: String) {
        guard let recipientBytes = hexToKotlinByteArray(recipientHex) else {
            log("❌ Invalid recipient hex: \(recipientHex)")
            return
        }
        let payloadBytes = stringToKotlinByteArray(message)
        let result = meshLink.send(recipient: recipientBytes, payload: payloadBytes)
        if result != nil {
            log("📤 Sent to \(String(recipientHex.prefix(8)))…: \(message)")
        } else {
            log("❌ Send failed")
        }
    }

    func broadcastMessage(_ message: String) {
        let payloadBytes = stringToKotlinByteArray(message)
        let _ = meshLink.broadcast(payload: payloadBytes, maxHops: 3)
        log("📡 Broadcast: \(message)")
    }

    // MARK: - Flow Collection

    private func startCollectingFlows() {
        collectFlow(meshLink.messages) { [weak self] (message: Message) in
            let senderHex = kotlinByteArrayToHex(message.senderId)
            let text = kotlinByteArrayToString(message.payload)
            self?.log("📨 From \(String(senderHex.prefix(8)))…: \(text)")
        }

        collectFlow(meshLink.peers) { [weak self] (event: PeerEvent) in
            if let found = event as? PeerEvent.Found {
                let peerHex = kotlinByteArrayToHex(found.peerId)
                self?.log("🔵 Peer: \(String(peerHex.prefix(8)))…")
                self?.peerCount += 1
                let now = Date()
                let existing = self?.discoveredPeers.first { $0.id == peerHex }
                let info = PeerInfo(
                    id: peerHex,
                    rssi: Int.random(in: -85 ... -35),
                    lastSeen: now,
                    firstSeen: existing?.firstSeen ?? now
                )
                if let idx = self?.discoveredPeers.firstIndex(where: { $0.id == peerHex }) {
                    self?.discoveredPeers[idx].rssi = info.rssi
                    self?.discoveredPeers[idx].lastSeen = info.lastSeen
                } else {
                    self?.discoveredPeers.append(info)
                }
            } else if let lost = event as? PeerEvent.Lost {
                let peerHex = kotlinByteArrayToHex(lost.peerId)
                self?.log("🔴 Lost: \(String(peerHex.prefix(8)))…")
                self?.peerCount = max(0, (self?.peerCount ?? 1) - 1)
                self?.discoveredPeers.removeAll { $0.id == peerHex }
            }
        }

        collectFlow(meshLink.transferFailures) { [weak self] (failure: TransferFailure) in
            self?.log("❌ Transfer failed: \(failure.messageId)")
        }

        collectFlow(meshLink.diagnosticEvents) { [weak self] (event: DiagnosticEvent) in
            guard let self else { return }
            self.diagnosticCounter += 1
            let entry = DiagnosticEntry(
                id: "\(event.monotonicMillis)-\(event.code.name)-\(self.diagnosticCounter)",
                code: event.code.name,
                severity: event.severity.name,
                timestamp: Self.formatTimestamp(event.monotonicMillis),
                payload: event.payload
            )
            if self.diagnosticEntries.count > 500 {
                self.diagnosticEntries.removeFirst()
            }
            self.diagnosticEntries.append(entry)
        }
    }

    // MARK: - Health

    func refreshHealth() {
        let snapshot = meshLink.meshHealth()
        connectedPeers = Int(snapshot.connectedPeers)
        reachablePeers = Int(snapshot.reachablePeers)
        powerMode = snapshot.powerMode
        bufferUsagePercent = Int(snapshot.bufferUtilizationPercent)
        activeTransfers = Int(snapshot.activeTransfers)
    }

    // MARK: - Logging

    private func log(_ entry: String) { logEntries.append(entry) }
    func clearLog() { logEntries.removeAll() }
    func clearDiagnostics() { diagnosticEntries.removeAll() }

    private static func formatTimestamp(_ monotonicMillis: Int64) -> String {
        let totalSeconds = monotonicMillis / 1000
        let minutes = (totalSeconds / 60) % 60
        let seconds = totalSeconds % 60
        let millis = monotonicMillis % 1000
        return String(format: "%02d:%02d.%03d", minutes, seconds, millis)
    }
}

// MARK: - Kotlin Flow → Swift Bridge

private class SwiftFlowCollector<T: AnyObject>: Kotlinx_coroutines_coreFlowCollector {
    let callback: (T) -> Void
    init(callback: @escaping (T) -> Void) { self.callback = callback }

    func emit(value: Any?) async throws {
        if let typedValue = value as? T {
            await MainActor.run { callback(typedValue) }
        }
    }
}

private func collectFlow<T: AnyObject>(
    _ flow: Kotlinx_coroutines_coreFlow,
    callback: @escaping (T) -> Void
) {
    Task.detached {
        let collector = SwiftFlowCollector<T>(callback: callback)
        do {
            try await flow.collect(collector: collector)
        } catch {
            print("Flow collection ended: \(error.localizedDescription)")
        }
    }
}

// MARK: - Byte Array Helpers

private func hexToKotlinByteArray(_ hex: String) -> KotlinByteArray? {
    let clean = hex.replacingOccurrences(of: " ", with: "")
    guard clean.count % 2 == 0 else { return nil }
    var bytes: [UInt8] = []
    var index = clean.startIndex
    while index < clean.endIndex {
        let next = clean.index(index, offsetBy: 2)
        guard let byte = UInt8(clean[index..<next], radix: 16) else { return nil }
        bytes.append(byte)
        index = next
    }
    let kb = KotlinByteArray(size: Int32(bytes.count))
    for (i, b) in bytes.enumerated() { kb.set(index: Int32(i), value: Int8(bitPattern: b)) }
    return kb
}

private func stringToKotlinByteArray(_ string: String) -> KotlinByteArray {
    let data = Array(string.utf8)
    let kb = KotlinByteArray(size: Int32(data.count))
    for (i, b) in data.enumerated() { kb.set(index: Int32(i), value: Int8(bitPattern: b)) }
    return kb
}

private func kotlinByteArrayToHex(_ bytes: KotlinByteArray) -> String {
    (0..<bytes.size).map { String(format: "%02x", UInt8(bitPattern: bytes.get(index: $0))) }.joined()
}

private func kotlinByteArrayToString(_ bytes: KotlinByteArray) -> String {
    let data = (0..<bytes.size).map { UInt8(bitPattern: bytes.get(index: $0)) }
    return String(bytes: data, encoding: .utf8) ?? "<invalid UTF-8>"
}

File: meshlink-sample/android/src/main/kotlin/io/meshlink/sample/MeshLinkViewModel.kt
package io.meshlink.sample

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import io.meshlink.MeshLink
import io.meshlink.config.MeshLinkConfig
import io.meshlink.crypto.CryptoProvider
import io.meshlink.diagnostics.DiagnosticEvent
import io.meshlink.diagnostics.MeshHealthSnapshot
import io.meshlink.diagnostics.Severity
import io.meshlink.model.PeerDetail
import io.meshlink.model.PeerEvent
import io.meshlink.transport.AndroidBleTransport
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** Lightweight descriptor of a discovered peer for the mesh visualizer. */
data class PeerInfo(
    val id: String,
    val rssi: Int,
    val lastSeen: Long,
    val firstSeen: Long,
)

class MeshLinkViewModel(application: Application) : AndroidViewModel(application) {

    private val transport = AndroidBleTransport(application)

    private var _currentConfig = MutableStateFlow(MeshLinkConfig.smallPayloadLowLatency { diagnosticsEnabled = true })
    /** Current [MeshLinkConfig] — observable for the Settings screen. */
    val currentConfig: StateFlow<MeshLinkConfig> = _currentConfig.asStateFlow()

    private val _currentPreset = MutableStateFlow(ConfigPreset.CHAT)
    /** Which config preset is active. */
    val currentPreset: StateFlow<ConfigPreset> = _currentPreset.asStateFlow()

    private val crypto = CryptoProvider()

    private var meshLink = MeshLink(transport, _currentConfig.value, crypto = crypto)

    private val _logs = MutableStateFlow<List<String>>(emptyList())
    val logs: StateFlow<List<String>> = _logs.asStateFlow()

    private val _health = MutableStateFlow(meshLink.meshHealth())
    val health: StateFlow<MeshHealthSnapshot> = _health.asStateFlow()

    private val _isRunning = MutableStateFlow(false)
    val isRunning: StateFlow<Boolean> = _isRunning.asStateFlow()

    private val _startTime = MutableStateFlow(0L)
    /** Epoch millis when the mesh was last started (0 = never). */
    val startTime: StateFlow<Long> = _startTime.asStateFlow()

    private val _discoveredPeers = MutableStateFlow<List<PeerInfo>>(emptyList())
    /** Live list of discovered peers for the mesh visualizer. */
    val discoveredPeers: StateFlow<List<PeerInfo>> = _discoveredPeers.asStateFlow()

    private val _diagnosticEvents = MutableStateFlow<List<DiagnosticEvent>>(emptyList())
    /** Rolling list of diagnostic events for the diagnostics screen. */
    val diagnosticEvents: StateFlow<List<DiagnosticEvent>> = _diagnosticEvents.asStateFlow()

    private val _severityFilter = MutableStateFlow<Severity?>(null)
    /** Current severity filter (null = show all). */
    val severityFilter: StateFlow<Severity?> = _severityFilter.asStateFlow()

    private val _selectedPeerId = MutableStateFlow<String?>(null)
    /** Currently selected peer ID in the mesh visualizer (null = none). */
    val selectedPeerId: StateFlow<String?> = _selectedPeerId.asStateFlow()

    /** Returns core library detail for the given peer, or null if unknown. */
    fun peerDetail(peerIdHex: String): PeerDetail? = meshLink.peerDetail(peerIdHex)

    fun selectPeer(peerId: String?) {
        _selectedPeerId.value = peerId
    }

    init {
        // Enable BLE debug logging for debug builds
        AndroidBleTransport.debugLogging = true

        log("🔧 Transport: AndroidBleTransport (peer: ${transport.localPeerId.toHex()})")
        collectFlows()
    }

    private fun collectFlows() {
        viewModelScope.launch {
            meshLink.messages.collect { message ->
                log("📨 Message from ${message.senderId.toHex()}: ${message.payload.decodeToString()}")
            }
        }
        viewModelScope.launch {
            meshLink.peers.collect { event ->
                when (event) {
                    is PeerEvent.Found -> {
                        val hexId = event.peerId.toHex()
                        log("🔵 Peer discovered: $hexId")
                        _discoveredPeers.update { peers ->
                            // Simulated RSSI — a real transport would supply the actual value.
                            val rssi = (-90..-40).random()
                            val now = System.currentTimeMillis()
                            val existing = peers.find { it.id == hexId }
                            peers.filterNot { it.id == hexId } + PeerInfo(
                                hexId,
                                rssi,
                                lastSeen = now,
                                firstSeen = existing?.firstSeen ?: now,
                            )
                        }
                    }
                    is PeerEvent.Lost -> {
                        val hexId = event.peerId.toHex()
                        log("🔴 Peer lost: $hexId")
                        _discoveredPeers.update { peers -> peers.filterNot { it.id == hexId } }
                    }
                }
            }
        }
        viewModelScope.launch {
            meshLink.meshHealthFlow.collect { snapshot ->
                _health.value = snapshot
                log("💓 Health update: peers=${snapshot.connectedPeers}, mode=${snapshot.powerMode.name}")
            }
        }
        viewModelScope.launch {
            meshLink.deliveryConfirmations.collect { uuid ->
                log("✅ Delivered: $uuid")
            }
        }
        viewModelScope.launch {
            meshLink.transferFailures.collect { failure ->
                log("❌ Transfer failed: ${failure.messageId}")
            }
        }
        viewModelScope.launch {
            meshLink.diagnosticEvents.collect { event ->
                _diagnosticEvents.update { events ->
                    (events + event).takeLast(500)
                }
            }
        }
    }

    fun startMesh() {
        meshLink.start()
            .onSuccess {
                _isRunning.value = true
                _startTime.value = System.currentTimeMillis()
                log("🟢 Mesh started")
            }
            .onFailure { e ->
                log("❌ Start failed: ${e.message}")
            }
    }

    fun stopMesh() {
        meshLink.stop()
        _isRunning.value = false
        _startTime.value = 0L
        _discoveredPeers.value = emptyList()
        _health.value = meshLink.meshHealth()
        log("🔴 Mesh stopped")
    }

    fun sendMessage(recipientId: String, message: String) {
        val recipientBytes = recipientId.hexToByteArray()
        val payloadBytes = message.encodeToByteArray()
        meshLink.send(recipientBytes, payloadBytes)
            .onSuccess { uuid ->
                log("📤 Sent (id=$uuid) to ${recipientId.take(12)}…")
            }
            .onFailure { e ->
                log("❌ Send failed: ${e.message}")
            }
    }

    fun broadcastMessage(message: String) {
        val payloadBytes = message.encodeToByteArray()
        meshLink.broadcast(payloadBytes, maxHops = _currentConfig.value.broadcastTTL)
            .onSuccess { uuid ->
                log("📡 Broadcast (id=$uuid): $message")
            }
            .onFailure { e ->
                log("❌ Broadcast failed: ${e.message}")
            }
    }

    /** Switch to a different config preset. Restarts the mesh if it was running. */
    fun applyPreset(preset: String) {
        val wasRunning = _isRunning.value
        if (wasRunning) meshLink.stop()

        val newConfig = when (preset) {
            "largePayloadHighThroughput" -> {
                _currentPreset.value = ConfigPreset.FILE_TRANSFER
                MeshLinkConfig.largePayloadHighThroughput { diagnosticsEnabled = true }
            }
            "minimalResourceUsage" -> {
                _currentPreset.value = ConfigPreset.POWER
                MeshLinkConfig.minimalResourceUsage { diagnosticsEnabled = true }
            }
            "minimalOverhead" -> {
                _currentPreset.value = ConfigPreset.SENSOR
                MeshLinkConfig.minimalOverhead { diagnosticsEnabled = true }
            }
            else -> {
                _currentPreset.value = ConfigPreset.CHAT
                MeshLinkConfig.smallPayloadLowLatency { diagnosticsEnabled = true }
            }
        }
        _currentConfig.value = newConfig
        meshLink = MeshLink(transport, newConfig, crypto = crypto)
        collectFlows()
        log("⚙️ Applied preset: $preset")

        if (wasRunning) startMesh()
    }

    /** Update the MTU in the current config. Restarts the mesh if it was running. */
    fun updateMtu(mtu: Int) {
        val wasRunning = _isRunning.value
        if (wasRunning) meshLink.stop()

        val newConfig = _currentConfig.value.copy(mtu = mtu)
        _currentConfig.value = newConfig
        meshLink = MeshLink(transport, newConfig, crypto = crypto)
        collectFlows()

        if (wasRunning) startMesh()
    }

    fun setSeverityFilter(severity: Severity?) {
        _severityFilter.value = severity
    }

    fun clearDiagnostics() {
        _diagnosticEvents.value = emptyList()
    }

    private fun log(entry: String) {
        _logs.update { it + entry }
    }

    override fun onCleared() {
        super.onCleared()
        meshLink.stop()
    }
}

private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }

private fun String.hexToByteArray(): ByteArray {
    check(length % 2 == 0) { "Hex string must have even length" }
    return chunked(2).map { it.toInt(16).toByte() }.toByteArray()
}
